### PR TITLE
Recompression policy

### PR DESCRIPTION
**Pull request by svenklemm**

### CODE REVIEW

File: tsl/test/expected/policy_recompression.out [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+(1 row)
+
+-- nothing to do yet
+CALL run_job(:JOB_ID);
+NOTICE:  no chunks for hypertable public.metrics that satisfy recompress chunk policy
+INSERT INTO metrics SELECT '2000-01-01';
+-- status should be 3
+SELECT * FROM _timescaledb_catalog.chunk WHERE id=1;
+ id | hypertable_id |      schema_name      |    table_name    | compressed_chunk_id | dropped | status 
+----+---------------+-----------------------+------------------+---------------------+---------+--------
```
Comment (gayyappan) [inline tsl/test/expected/policy_recompression.out:42]: looks like something is not right here.

File: tsl/test/expected/policy_recompression.out [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  1 |             1 | _timescaledb_internal | _hyper_1_1_chunk |                     | f       |      0
+(1 row)
+
+CALL run_job(:JOB_ID);
+NOTICE:  no chunks for hypertable public.metrics that satisfy recompress chunk policy
```
Comment (gayyappan) [inline tsl/test/expected/policy_recompression.out:47]: The job has nothing to do since status is still 1?

File: src/jsonb_utils.c [hunk 3]
```diff
@@ -171,6 +171,24 @@ ts_jsonb_get_time_field(const Jsonb *jsonb, const char *key, bool *field_found)
 	return DatumGetTimestampTz(time_datum);
 }
 
+bool
+ts_jsonb_get_bool_field(const Jsonb *json, const char *key, bool *field_found)
```
Comment (gayyappan) [inline src/jsonb_utils.c:175]: This is unused now. Will this be used by compression policy?

File: tsl/src/bgw_policy/job.c [hunk 4]
```diff
@@ -551,6 +575,61 @@ policy_compression_read_and_validate_config(Jsonb *config, PolicyCompressionData
 	}
```
Comment (gayyappan) [inline tsl/src/bgw_policy/job.c:575]: don't we have to check for boolean flag here?

File: tsl/test/expected/policy_recompression.out [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+----+---------------+-----------------------+------------------+---------------------+---------+--------
+  1 |             1 | _timescaledb_internal | _hyper_1_1_chunk |                     | f       |      0
+(1 row)
+
+SELECT delete_job(:JOB_ID);
+ delete_job 
+------------
+ 
+(1 row)
+
```
Comment (gayyappan) [inline tsl/test/expected/policy_recompression.out:60]: Also need a test with compression policy and recompress parameter = false.

**Review by gayyappan — COMMENTED** [review]
something not quite right in the test case.

File: src/dimension_slice.c [hunk 6]
```diff
@@ -931,25 +931,34 @@ ts_dimension_slice_oldest_valid_chunk_for_reorder(int32 job_id, int32 dimension_
...
 
+typedef struct FindChunkCompress
+{
+	int32 chunk_id;
+	bool compress;
+	bool recompress;
+} FindChunkCompress;
+
 static ScanTupleResult
 dimension_slice_check_is_chunk_uncompressed_tuple_found(TupleInfo *ti, void *data)
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply src/dimension_slice.c:942]: if we change function signature then it wouldnt match the callback signature anymore

File: tsl/src/bgw_policy/compression_api.c [hunk 7]
```diff
@@ -37,8 +37,20 @@
...
 #define CONFIG_KEY_HYPERTABLE_ID "hypertable_id"
 #define CONFIG_KEY_COMPRESS_AFTER "compress_after"
+#define CONFIG_KEY_RECOMPRESS_AFTER "recompress_after"
+#define CONFIG_KEY_RECOMPRESS "recompress"
+
+bool
+policy_compression_get_recompress(const Jsonb *config)
+{
+	bool found;
+	bool recompress = ts_jsonb_get_int32_field(config, CONFIG_KEY_RECOMPRESS, &found);
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply tsl/src/bgw_policy/compression_api.c]: fixed

File: src/jsonb_utils.c (hunk 3 above)
Comment (svenklemm) [reply src/jsonb_utils.c:175]: yes

File: tsl/test/expected/policy_recompression.out [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+SELECT create_hypertable('metrics','time');
+  create_hypertable   
+----------------------
+ (1,public,metrics,t)
+(1 row)
+
+ALTER TABLE metrics SET (timescaledb.compress);
+-- create chunk with some data and compress
+INSERT INTO metrics SELECT '2000-01-01' FROM generate_series(1,10);
+-- create custom recompression job
```
Comment (gayyappan) [inline tsl/test/expected/policy_recompression.out:14]:
```suggestion
-- create custom compression job that has no special config value set
```

**Review by gayyappan — APPROVED** [review]

**Pull request merged**